addons/project/controllers/portal.py

```diff
@@ -70,13 +107,54 @@ def portal_my_projects(self, page=1, date_begin=None, date_end=None, sortby=None
         })
         return request.render("project.portal_my_projects", values)
 
-    @http.route(['/my/project/<int:project_id>'], type='http', auth="user", website=True)
-    def portal_my_project(self, project_id=None, **kw):
-        project = request.env['project.project'].browse(project_id)
-        vals = {'project': project}
-        history = request.session.get('my_projects_history', [])
-        vals.update(get_records_pager(history, project))
-        return request.render("project.portal_my_project", vals)
+    @http.route(['/my/project/<int:project_id>'], type='http', auth="public", website=True)
+    def portal_my_project(self, project_id=None, access_token=None, **kw):
+        try:
+            project_sudo = self._project_check_access(project_id, access_token)
+        except AccessError:
+            return request.redirect('/my')
+
+        values = self._project_get_page_view_values(project_sudo, access_token, **kw)
+        return request.render("project.portal_my_project", values)
+
+
+
+    # ------------------------------------------------------------
+    # My Task
+    # ------------------------------------------------------------
+    def _task_check_access(self, task_id, access_token=None):
+        task = request.env['project.task'].browse([task_id])
+        task_sudo = task.sudo()
+        try:
+            task.check_access_rights('read')
+            task.check_access_rule('read')
+        except AccessError:
+            if not access_token or not consteq(task_sudo.access_token, access_token):
+                raise
+        return task_sudo
+
+    def _task_get_page_view_values(self, task, access_token, **kwargs):
+        values = {
+            'page_name': 'task',
+            'task': task,
+            'user': request.env.user
+        }
+        if access_token:
```

Comment:
There are several times the same code (no breadcrum, access token, taking error / warning / success from kwargs). Shouldn't we try to factorize it a bit so that we keep only application-specific code ?

Reply:
Yep ! It's quite the same thing that @jem-odoo noticed. A lot of code in those controllers is the same for all models.